megatron/core/inference/text_generation_controllers/text_generation_controller.py

```diff
@@ -722,6 +736,54 @@ def _dynamic_step_calculate_log_probs(self, logits: Tensor) -> Optional[Tensor]:
         )
         return ret
 
+    def _dynamic_step_calculate_top_n_logprobs(self, logits: Tensor) -> Optional[Dict[int, List[Tuple[Tensor, Tensor]]]]:
+        """Calculate top-n log probs from logits for dynamic batching.
+        
+        Returns:
+            A dictionary mapping request_idx to list of (top_n_logprobs, top_n_indices) tuples.
+            Each tuple in the list represents one token position.
+        """
+        context = self.inference_wrapped_model.inference_context
+        materialize_only_last_token_logits = context.materialize_only_last_token_logits
+
+        active_request_count = context.total_request_count - context.paused_request_count
+
+        # Get logits - handle both decode (last token only) and prefill (all tokens) modes
+        if materialize_only_last_token_logits:
+            # Decode mode: only last token logits
+            # Shape: [1, active_request_count, vocab_size] -> [active_request_count, vocab_size]
+            last_token_logits = logits.squeeze(0)
+            # Compute log probabilities
+            log_probs = torch.nn.functional.log_softmax(last_token_logits, dim=-1).to(torch.float32)
+            
+            # For each request, get top-n logprobs based on its top_n_logprobs value
+            top_n_results = {}
+            for req_idx in range(active_request_count):
+                top_n = int(self.top_n_logprobs_cuda[req_idx].item())
+                if top_n > 0:
+                    # Get top-n logprobs and indices for this request (single token)
+                    top_n_logits = torch.topk(log_probs[req_idx], k=min(top_n, log_probs.size(-1)))
+                    top_n_results[req_idx] = [(top_n_logits.values, top_n_logits.indices)]
+        else:
+            # Prefill/non-decode mode: all tokens
+            # Shape: [1, num_tokens, vocab_size]
+            # Need to extract per-request logits using last_token_logits helper
+            last_token_logits = context.last_token_logits(logits)  # Shape: [active_request_count, vocab_size]
+            
+            # Compute log probabilities for last tokens
```

Comment:
This misses logprobs for prefill.

This is why I would reuse the implementation in `dynamic_context.py`, under `calculate_log_probs`. That accounts for prefill logprobs, and the logic to do so is non-trivial.